core/location/cache.go

```diff
@@ -17,26 +17,76 @@
 
 package location
 
-type cache struct {
+import (
+	"sync"
+	"time"
+
+	log "github.com/cihub/seelog"
+	"github.com/mysteriumnetwork/node/core/connection"
+)
+
+const locationCacheLogPrefix = "[location-cache]"
+
+// Cache allows us to cache location resolution
+type Cache struct {
+	lastFetched      time.Time
 	locationDetector Resolver
 	location         Location
+	expiry           time.Duration
+	lock             sync.Mutex
+}
+
+// NewCache returns a new instance of location cache
+func NewCache(resolver Resolver, expiry time.Duration) *Cache {
+	return &Cache{
+		locationDetector: resolver,
+		expiry:           expiry,
+	}
+}
+
+func (c *Cache) needsRefresh() bool {
+	return c.lastFetched.IsZero() || c.lastFetched.After(time.Now().Add(-c.expiry))
 }
 
-// NewLocationCache constructs Cache
-func NewLocationCache(locationDetector Resolver) Cache {
-	return &cache{
-		locationDetector: locationDetector,
+func (c *Cache) fetchAndSave() (Location, error) {
+	loc, err := c.locationDetector.DetectLocation()
+
+	// on successful fetch save the values for further use
+	if err == nil {
+		c.location = loc
+		c.lastFetched = time.Now()
 	}
+
+	return loc, err
 }
 
-// Gets location from cache
-func (lc *cache) Get() Location {
-	return lc.location
+// DetectLocation returns location from cache, or fetches it if needed
+func (c *Cache) DetectLocation() (Location, error) {
+	c.lock.Lock()
```

Comment:
Maybe lock should go to the `fetchAndSave()` and `needsRefresh`?
I see that it's not called outside locks, but one day someone will change it and break it.

Reply:
I think that the private methods should not be locked. The lock is purposefully put on the public methods, as I want to ensure that no double update occurs. Imagine the lock moved onto the private methods. Two routines call DetectLocation. First one checks if we need to refresh. Finds out we do. The lock is released, we move further. The second routine then kicks in, also performs the same check, finds out we still need to refresh. Both of them proceed to update the state, where only a single update is needed.